splash/html_element.py

```diff
@@ -0,0 +1,222 @@
+        });
+    }
+    if (type === 'radio') {
+        return field.checked ? field.value : null;
+    }
+    if (type === 'checkbox') {
+        return field.checked;
+    }
+    return field.value || '';
+}
+"""
+
+# Casper.JS
+FORM_VALUES_JS = """
+function(form, getField) {
+    var values = {}, checked = {};
+
+    [].forEach.call(form.elements, function(elm) {
+        var name = elm.getAttribute('name');
+        var value = getField(elm);
+        var multi = !!value && elm.hasAttribute('type') &&
+                    elm.type === 'checkbox' ? elm.value : value;
+        if (!!name && value !== null && !(elm.type === 'checkbox' && value === false)) {
+            if (typeof values[name] === "undefined") {
+                values[name] = value;
+                checked[name] = multi;
+            } else {
+                if (!Array.isArray(values[name])) {
+                    values[name] = [checked[name]];
+                }
+                values[name].push(multi);
+            }
+        }
+    });
+    return values;
+}
+"""
+
+
+class HTMLElement(object):
+    def __init__(self, tab, storage, node):
+        if node == "":
+            raise RuntimeError("Cannot find the requested element")
+
+        self.tab = tab
+        self.storage = storage
+        self.id = node["id"]
+        self.element_js = self.get_element_js()
+
+    def get_element_js(self):
+        return "window['%s']['%s']" % (self.storage.name, self.id)
+
+    def assert_element_exists(self):
+        if not self.exists():
+            raise RuntimeError("Element no longer exists in DOM")
+
+    def exists(self):
+        try:
+            exists = self.tab.evaljs("document.contains(%s)" % self.element_js)
+            return bool(exists)
+        except JsError:
+            return False
+
+    def node_property(self, property_name):
+        self.assert_element_exists()
+        return self.tab.evaljs("%s[%s]" % (self.element_js, escape_js(property_name)))
+
+    def node_method(self, method_name):
+        self.assert_element_exists()
+        return self.tab.evaljs("%s[%s]()" % (self.element_js, escape_js(method_name)))
+
+    def mouse_click(self, button="left"):
+        self.assert_element_exists()
+        dimensions = self.tab.evaljs(
+            DIMENSIONS_JS_FUNC % self.element_js,
+        )
+
+        self.tab.mouse_click(dimensions["x"], dimensions["y"], button)
+
+    def mouse_hover(self):
+        self.assert_element_exists()
+        dimensions = self.tab.evaljs(
+            DIMENSIONS_JS_FUNC % self.element_js,
+        )
+
+        self.tab.mouse_hover(dimensions["x"], dimensions["y"])
+
+    def get_styles(self):
+        self.assert_element_exists()
+        return self.tab.evaljs("getComputedStyle(%s)" % self.element_js, False, False)
+
+    def get_bounds(self):
+        self.assert_element_exists()
+        return self.tab.evaljs("%s.getClientRects()[0]" % self.element_js, result_protection=False)
```

Comment:
This getClientRects/getBoundingClientRect debate makes me wonder:
- if we want to expose getBoundingClientRect to the users
- if we want to expose getClientRects returning an array to the users
- if we want to capture (png) all client rects or the bounding client rect to the users